File: eticket_management/e_ticket_management/report/ticket_sold_report/ticket_sold_report.py
# Copyright (c) 2022, Frappe Technologies and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import date_diff

def execute(filters=None):
	data=[]
	if filters.parent_row_group != "None":
		data = get_report_group_data(filters)
	else:
		data = get_data(filters)
	validate(filters)
	return get_columns(filters), data

def get_filters(filters):
	data = " transaction_date between '{}' AND '{}'".format(filters.start_date,filters.end_date)
	data = data + "{}={}".format()
	return data

def get_columns(filters):
	columns = []
	columns.append({'fieldname':"group_field",'label':"{}".format(filters.row_group),'fieldtype':'Data','align':'left','width':200})
	if filters.column_group != "None":
		fields = get_fields(filters)
		for f in fields:
				columns.append({'fieldname':"qty_"+f['fieldname'],'label':"Total QTY " + f["label"],'align':'right','fieldtype':'Data'})
				columns.append({'fieldname':f['fieldname'],'label':"Total Amt " + f["label"],'align':'right','fieldtype':'Currency'})
	columns.append({'fieldname':"total_amount",'label':"Total Amount",'fieldtype':'Currency','align':'right','width':110})
	columns.append({'fieldname':"total_ticket",'label':"Total Ticket",'fieldtype':'Data','align':'center','width':100})
	return columns

def get_report_group_data(filters):
	parent = get_data(filters,0,get_row_group(filters.parent_row_group))
	data=[]
	for p in parent:
		p["is_group"] = 1
		data.append(p)
		children = get_data(filters, 1,get_row_group(filters.parent_row_group),parent["group_field"])
		for c in children:data.append(c)
	return data

def get_data(filters,indent=0,parent_row_group=""):
	row_group = ""
	if parent_row_group != "":
		row_group = parent_row_group
	else:
		row_group = get_row_group(filters.row_group)
	sql = "SELECT {} as group_field,{} as indent,".format(row_group,indent)
	data=[]
	if filters.column_group != "None":
		fields = get_fields(filters)
		for f in fields:
			sql = sql +	"""
			SUM(if(transaction_date between '{0}' AND '{1}',{2},0)) as '{3}',
			sum(if(transaction_date between '{0}' AND '{1}',1,0)) as 'qty_{3}',
			""".format(f["start_date"],f["end_date"],"price",f["fieldname"])
	sql = sql + """
				sum(price) total_amount,
				count(name) total_ticket
				FROM `tabPOS Ticket` a
					WHERE {0}
				group by {1}
			""".format(get_filters(filters),row_group)
	frappe.msgprint(sql)
	data = frappe.db.sql(sql,as_dict=1)
	return data

def get_list(filters,name):
	data = ','.join("'{0}'".format(x) for x in filters.get(name))
	return data

def get_row_group(row_group):
	data= ""
	if(row_group) == "Item": data = "item_name"
	if(row_group) == "Sale Invoice": data = "pos_invoice"
	if(row_group) == "Cashier": data = "a.pos_username"
	return data

def validate(filters):

	if filters.start_date and filters.end_date:
		if filters.start_date > filters.end_date:
			frappe.throw("The 'Start Date' ({}) must be before the 'End Date' ({})".format(filters.start_date, filters.end_date))

	if filters.column_group=="Daily":
		n = date_diff(filters.end_date, filters.start_date)
		if n>30:
			frappe.throw("Date range cannot greater than 30 days")

def get_fields(filters):
	sql=""
	if filters.column_group=="Daily":
		sql = """
			select 
				concat(date_format(date,'%d_%m')) as fieldname, 
				date_format(date,'%d') as label ,
				min(date) as start_date,
				max(date) as end_date
			from `tabDates` 
			where date between '{}' and '{}'
			group by
				concat(date_format(date,'%d_%m')) , 
				date_format(date,'%d')  	
		""".format(filters.start_date, filters.end_date)
	elif filters.column_group =="Monthly":
		sql = """
			select 
				concat(date_format(date,'%m_%Y')) as fieldname, 
				date_format(date,'%b %y') as label ,
				min(date) as start_date,
				max(date) as end_date
			from `tabDates` 
			where date between '{}' and '{}'
			group by
				concat(date_format(date,'%m_%Y')) , 
				date_format(date,'%b %y')  	
		""".format(filters.start_date, filters.end_date)
	elif filters.column_group=="Yearly":
		sql = """
			select 
				concat(date_format(date,'%Y')) as fieldname, 
				date_format(date,'%Y') as label ,
				min(date) as start_date,
				max(date) as end_date
			from `tabDates` 
			where date between '{}' and '{}'
			group by
				concat(date_format(date,'%Y')),
				date_format(date,'%Y')
		""".format(filters.start_date, filters.end_date)
	fields = frappe.db.sql(sql,as_dict=1)
	return fields

File: eticket_management/e_ticket_management/report/ticket_sold_report/ticket_sold_report.js
// Copyright (c) 2022, Frappe Technologies and contributors
// For license information, please see license.txt
/* eslint-disable */
frappe.query_reports["Ticket Sold Report"] = {
	"filters": [
		{
			fieldname: "company",
			label: "Company",
			fieldtype: "Link",
			options:"Company",
		},
		{
			"fieldname":"filter_based_on",
			"label": __("Filter Based On"),
			"fieldtype": "Select",
			"options": ["Fiscal Year", "Date Range"],
			"default": ["Fiscal Year"],
			"reqd": 1,
			on_change: function() {
				let filter_based_on = frappe.query_report.get_filter_value('filter_based_on');
				frappe.query_report.toggle_filter_display('from_fiscal_year', filter_based_on === 'Date Range');
				frappe.query_report.toggle_filter_display('start_date', filter_based_on === 'Fiscal Year');
				frappe.query_report.toggle_filter_display('end_date', filter_based_on === 'Fiscal Year');

				frappe.query_report.refresh();
			}
		},
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			default:frappe.datetime.get_today(),
			"hidden": 1,
			"reqd": 1
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			default:frappe.datetime.get_today(),
			"hidden": 1,
			"reqd": 1
		},
		{
			"fieldname":"from_fiscal_year",
			"label": __("Start Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"default": frappe.defaults.get_user_default("fiscal_year"),
			"reqd": 1,
			on_change: () => {
				frappe.model.with_doc("Fiscal Year", frappe.query_report.get_filter_value('from_fiscal_year'), function(r) {
					let year_start_date = frappe.model.get_value("Fiscal Year", frappe.query_report.get_filter_value('from_fiscal_year'), "year_start_date");
					frappe.query_report.set_filter_value({
						period_start_date: year_start_date
					});
				});
			}
		},

		{
			"fieldname": "price_list",
			"label": __("Sale Type"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Price List', txt,{"selling":1});
			}
		},
		{
			"fieldname": "customer_group",
			"label": __("Customer Group"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				
				return frappe.db.get_link_options('Customer Group', txt,{"is_group":0});
			}
		},
		{
			fieldname: "market_segment",
			label: "Market Segment",
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Market Segment', txt);
			}
		},
		{
			fieldname: "business_source",
			label: "Business Source",
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Lead Source', txt);
			}
		},
		{
			"fieldname": "parent_row_group",
			"label": __("Parent Group By"),
			"fieldtype": "Select",
			"options": "\nCategory\nProduct Group\nBrand\nCompany\nPrice List\nCustomer\nCustomer Group\nMembership\nTerritory\nDate\n\Month\nYear\nSale Invoice\nMarket Segment\nMarketing Segment Type\nBusiness Source\nBusiness Source Type",
			
		},
		{
			"fieldname": "row_group",
			"label": __("Row Group By"),
			"fieldtype": "Select",
			"options": "Product\nCategory\nProduct Group\nBrand\nCompany\nPrice List\nCustomer\nCustomer Group\nMembership\nTerritory\nDate\n\Month\nYear\nSale Invoice\nMarket Segment\nMarketing Segment Type\nBusiness Source\nBusiness Source Type",
			"default":"Category"
		},
		{
			"fieldname": "column_group",
			"label": __("Column Group By"),
			"fieldtype": "Select",
			"options": "Column Group By\nDaily\nWeekly\nMonthly\nQuarterly\nHalf Yearly\nYearly",
			"default":"Column Group By"
		},
		{
			"fieldname": "hide_columns",
			"label": __("Hide Columns"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return [
					{"value":"Amount","description":"Amount"},
					{"value":"Quantity","description":"Quantity"},
					{"value":"Sub Total","description":"Sub Total"},
					 
					{"value":"Total Discount","description":"Total Discount"},
				]
			},
			"default":"All"
		},
		{
			"fieldname": "chart_type",
			"label": __("Chart Type"),
			"fieldtype": "Select",
			"options": "None\nbar\nline\npie",
			"default":"bar"
		}
	],
};

 